src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardDictionaryRemoveByContainsKey.cs

```diff
@@ -0,0 +1,119 @@
+
+                if (invocationOperation.TargetMethod.Name != "ContainsKey")
+                    return;
+
+                if (invocationOperation.Parent is not IConditionalOperation parentConditionalOperation)
+                    return;
+
+                // we only want to report this diagnostic if the Contains/Remove pair is all there is
+
+                if (parentConditionalOperation.WhenFalse == null &&
+                    parentConditionalOperation.WhenTrue.Children.HasExactly(1))
+                {
+                    var properties = ImmutableDictionary.CreateBuilder<string, string>();
+                    properties[ConditionalOperation] = CreateLocationInfo(parentConditionalOperation.Syntax);
+
+                    switch (parentConditionalOperation.WhenTrue.Children.First())
+                    {
+                        case IInvocationOperation childInvocationOperation:
+                            if (childInvocationOperation.TargetMethod.Name == "Remove")
+                            {
+                                properties[ChildStatementOperation] = CreateLocationInfo(childInvocationOperation.Syntax.Parent);
+
+                                context.ReportDiagnostic(Diagnostic.Create(Rule, invocationOperation.Syntax.GetLocation(), properties.ToImmutable()));
+                            }
+
+                            break;
+                        case IExpressionStatementOperation childStatementOperation:
+                            // if the if statement contains a block, only proceed if that block contains a single statement
+
+                            if (childStatementOperation.Children.HasExactly(1) &&
+                                childStatementOperation.Children.First() is CodeAnalysis.Operations.IInvocationOperation nestedInvocationOperation &&
+                                nestedInvocationOperation.TargetMethod.Name == "Remove")
+                            {
+                                properties[ChildStatementOperation] = CreateLocationInfo(nestedInvocationOperation.Syntax.Parent);
+
+                                context.ReportDiagnostic(Diagnostic.Create(Rule, invocationOperation.Syntax.GetLocation(), properties.ToImmutable()));
+                            }
+
+                            break;
+                        default:
+                            break;
+                    }
+                }
+            }
+        }
+
+        private static LocalizableString CreateResource(string resourceName)
+            => new LocalizableResourceString(resourceName, Resx.ResourceManager, typeof(Resx));
+
+        private static string CreateLocationInfo(SyntaxNode syntax)
+        {
+            // see DiagnosticDescriptorCreationAnalyzer
+
+            var location = syntax.GetLocation();
+            var span = location.SourceSpan;
+
+            return $"{span.Start}{AdditionalDocumentLocationInfoSeparator}{span.Length}";
```

Comment:
Is this separator meant to be only working for this analyzer or to be some kind of shared info? If the latter, might be worth extracting this to the shared diagnostic helper.

Reply:
>Is this separator meant to be only working for this analyzer or to be some kind of shared info?

The info is shared with the fixer (where it is consumed), which is why it isn't `private`.

>If the latter, might be worth extracting this to the shared diagnostic helper.

I was hoping there's already some kind of general pattern or helper for sharing a `TextSpan` between an analyzer and a fixer. Or is there a better approach at this?

Reply:
@mavasani I do recall you did something somewhere, sorry my memory is playing with me these days...